### PR TITLE
Use first and last, rather than front, back, top etc

### PR DESCRIPTION
**Pull request by krish8484**
Changes made in docs,src and test folders

### CODE REVIEW

File: src/circ_deque.jl [hunk 1]
```diff
@@ -72,7 +72,7 @@ end
 """
     pushfirst!(D::CircularDeque, v)
 
-Add an element to the front
+Add an element to the first
```
Comment (oxinabox) [inline src/circ_deque.jl]:
```suggestion
Add an element to the front.
```

File: src/circ_deque.jl [hunk 2]
```diff
@@ -33,21 +33,21 @@ end
...
     @boundscheck D.n > 0 || throw(BoundsError())
     D.buffer[D.first]
 end
 
 """
-    back(D::CircularDeque)
+    last(D::CircularDeque)
 
-Add an element to the back
+Add an element to the last
```
Comment (oxinabox) [inline src/circ_deque.jl]:
```suggestion
Add an element to the back.
```

File: src/circ_deque.jl [hunk 3]
```diff
@@ -86,10 +86,10 @@ end
 """
     pushfirst!(D::CircularDeque)
 
-Remove the element at the front
+Remove the element at the first
```
Comment (oxinabox) [inline src/circ_deque.jl]:
```suggestion
Remove the element at the front.
```

File: docs/src/circ_deque.md [hunk 4]
```diff
@@ -12,19 +12,19 @@ isempty(a)          # test whether the deque is empty
...
+first(a)            # get the element at the first
+last(a)             # get the element at the last
 eltype(a)           # return type of items
 ```
 
 *Note:* Julia's `Vector` type also provides this interface, and thus can
 be used as a deque. However, the `CircularDeque` type in this package is
 implemented as a circular buffer, and thus avoids copying elements when
-modifications are made to the front of the vector.
+modifications are made to the first of the vector.
```
Comment (oxinabox) [inline docs/src/circ_deque.md]:
```suggestion
modifications are made to the front of the vector.
```

File: docs/src/circ_deque.md [hunk 5]
```diff
@@ -12,19 +12,19 @@ isempty(a)          # test whether the deque is empty
...
-pushfirst!(a, 20)   # add an element to the front
-popfirst!(a)        # remove an element from the front
-front(a)            # get the element at the front
-back(a)             # get the element at the back
+push!(a, 10)        # add an element to the last
+pop!(a)             # remove an element from the last
+pushfirst!(a, 20)   # add an element to the first
+popfirst!(a)        # remove an element from the first
+first(a)            # get the element at the first
+last(a)             # get the element at the last
```
Comment (oxinabox) [inline docs/src/circ_deque.md]:
```suggestion
last(a)             # get the element at the back
```

File: docs/src/circ_deque.md [hunk 6]
```diff
@@ -12,19 +12,19 @@ isempty(a)          # test whether the deque is empty
...
-pop!(a)             # remove an element from the back
-pushfirst!(a, 20)   # add an element to the front
-popfirst!(a)        # remove an element from the front
-front(a)            # get the element at the front
-back(a)             # get the element at the back
+push!(a, 10)        # add an element to the last
+pop!(a)             # remove an element from the last
+pushfirst!(a, 20)   # add an element to the first
+popfirst!(a)        # remove an element from the first
+first(a)            # get the element at the first
```
Comment (oxinabox) [inline docs/src/circ_deque.md]:
```suggestion
first(a)            # get the element at the front
```

File: docs/src/circ_deque.md [hunk 7]
```diff
@@ -12,19 +12,19 @@ isempty(a)          # test whether the deque is empty
...
 length(a)           # get the number of elements currently in the deque
-push!(a, 10)        # add an element to the back
-pop!(a)             # remove an element from the back
-pushfirst!(a, 20)   # add an element to the front
-popfirst!(a)        # remove an element from the front
-front(a)            # get the element at the front
-back(a)             # get the element at the back
+push!(a, 10)        # add an element to the last
+pop!(a)             # remove an element from the last
+pushfirst!(a, 20)   # add an element to the first
```
Comment (oxinabox) [inline docs/src/circ_deque.md]:
```suggestion
pushfirst!(a, 20)   # add an element to the front
```

File: docs/src/circ_deque.md [hunk 8]
```diff
@@ -12,19 +12,19 @@ isempty(a)          # test whether the deque is empty
...
 capacity(a)         # return capacity
 length(a)           # get the number of elements currently in the deque
-push!(a, 10)        # add an element to the back
-pop!(a)             # remove an element from the back
-pushfirst!(a, 20)   # add an element to the front
-popfirst!(a)        # remove an element from the front
-front(a)            # get the element at the front
-back(a)             # get the element at the back
+push!(a, 10)        # add an element to the last
+pop!(a)             # remove an element from the last
```
Comment (oxinabox) [inline docs/src/circ_deque.md]:
```suggestion
pop!(a)             # remove an element from the back
```

File: docs/src/circ_deque.md [hunk 9]
```diff
@@ -12,19 +12,19 @@ isempty(a)          # test whether the deque is empty
 empty!(a)           # reset the deque
 capacity(a)         # return capacity
 length(a)           # get the number of elements currently in the deque
-push!(a, 10)        # add an element to the back
-pop!(a)             # remove an element from the back
-pushfirst!(a, 20)   # add an element to the front
-popfirst!(a)        # remove an element from the front
-front(a)            # get the element at the front
-back(a)             # get the element at the back
+push!(a, 10)        # add an element to the last
```
Comment (oxinabox) [inline docs/src/circ_deque.md]:
```suggestion
push!(a, 10)        # add an element to the front
```

File: src/circ_deque.jl (hunk 1 above)
Comment (krish8484) [reply src/circ_deque.jl]: Correct the English?

**Comment by krish8484** [comment]
@oxinabox will the @deprecate come in the src folder of each data structure namely; queue.jl, dequeue.jl and circ-queue.jl, right?

**Comment by krish8484** [comment]
@oxinabox there was a back function in stack as well, which was causing the build tests to fail, replaced that, now it's fine

**Comment by oxinabox** [comment]
If you add a `src/deprecations.jl` file I think that would be best.
The deprecation belongs to all of them but you only need to implement it once.

File: Project.toml [hunk 10]
```diff
@@ -1,6 +1,6 @@
 name = "DataStructures"
 uuid = "864edb3b-99cc-5e75-8d2d-829cb0a9cfe8"
-version = "0.17.6"
+version = "0.18.0"
```
Comment (oxinabox) [inline Project.toml]:
```suggestion
version = "0.17.7"
```

This is not a breaking change.
Only breaking changes increment the first nonzero

File: src/circ_deque.jl [hunk 11]
```diff
@@ -33,21 +33,22 @@ end
 Base.isempty(D::CircularDeque) = D.n == 0
 
 """
-    front(D::CircularDeque)
+    first(D::CircularDeque)
 
-Add an element to the front
+Add an element to the first
```
Comment (oxinabox) [inline src/circ_deque.jl]:
```suggestion
Add an element to the front.
```

File: src/stack.jl [hunk 12]
```diff
@@ -40,7 +40,7 @@ Base.eltype(::Type{Stack{T}}) where T = T
 
 Get a top item from a stack
 """
-top(s::Stack) = back(s.store)
+top(s::Stack) = last(s.store)
```
Comment (oxinabox) [inline src/stack.jl]: we should replace and deprecate this too.

```suggestion
first(s::Stack) = last(s.store)
```

**Comment by oxinabox** [comment]
Remaining:

- [ ] update docstring and tests and deprecation for `top`
- [ ] `include(...)` the deprecation file in the main `src/DataStructures.jl`
- [ ] remove the exports of the deprecated names. (`@deprecate` does export automatically)

**Comment by oxinabox** [comment]
I changed the name back.
Note that we have automatic release notes that say what was in a release.
https://github.com/JuliaCollections/DataStructures.jl/releases

File: src/circ_deque.jl [hunk 13]
```diff
@@ -33,21 +33,22 @@ end
 Base.isempty(D::CircularDeque) = D.n == 0
 
 """
-    front(D::CircularDeque)
+    first(D::CircularDeque)
 
-Add an element to the front
+Add an element to the front.
```
Comment (oxinabox) [inline src/circ_deque.jl]:
```suggestion
Get the item at the front of the queue.
```

File: src/circ_deque.jl [hunk 14]
```diff
@@ -33,21 +33,22 @@ end
...
 """
-@inline function front(D::CircularDeque)
+#@deprecate front(x) first(x)
```
Comment (oxinabox) [inline src/circ_deque.jl]:
```suggestion
```

File: src/circ_deque.jl [hunk 15]
```diff
@@ -33,21 +33,22 @@ end
...
     @boundscheck D.n > 0 || throw(BoundsError())
     D.buffer[D.first]
 end
 
 """
-    back(D::CircularDeque)
+    last(D::CircularDeque)
 
-Add an element to the back
+Add an element to the back.
```
Comment (oxinabox) [inline src/circ_deque.jl]:
```suggestion
Get the item from the back of the queue.
```

File: src/circ_deque.jl [hunk 16]
```diff
@@ -86,10 +87,10 @@ end
 """
     pushfirst!(D::CircularDeque)
```
Comment (oxinabox) [inline src/circ_deque.jl]:
```suggestion
    popfirst!(D::CircularDeque)
```

**Comment by krish8484** [comment]
I reckon, that should be it! @oxinabox

**Comment by krish8484** [comment]
I reckon, we are done!

**Pull request merged**